[package] Add `allow_empty` flag to mock and extern

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53233 [package] Add `deny` method to `PackageExporter`
* **#53232 [package] Add `allow_empty` flag to mock and extern**

**Summary**
This commit adds an optional `allow_empty` argument to
`PackageExporter.mock` and `PackageExporter.extern` that allows certain
patterns for mocked modules and extern modules to be marked ones that
*must* be matched during the packaging process. If a mock or extern
module with `allow_empty=False` is not matched while packaging, an error
is thrown.

**Test Plan**
This commit adds two new test cases to `PackagingTest`,
`test_extern_glob_allow_empty` and `test_mock_glob_allow_empty` that
test this new flag. Existing tests already tests `allow_empty=True`.

**Fixes**
This commit fixes #53217.

Differential Revision: [D26834011](https://our.internmc.facebook.com/intern/diff/D26834011)

Commit:
[package] Add `allow_empty` flag to mock and extern

**Summary**
This commit adds an optional `allow_empty` argument to
`PackageExporter.mock` and `PackageExporter.extern` that allows certain
patterns for mocked modules and extern modules to be marked ones that
*must* be matched during the packaging process. If a mock or extern
module with `allow_empty=False` is not matched while packaging, an error
is thrown.

**Test Plan**
This commit adds two new test cases to `PackagingTest`,
`test_extern_glob_allow_empty` and `test_mock_glob_allow_empty` that
test this new flag. Existing tests already tests `allow_empty=True`.

**Fixes**
This commit fixes #53217.

[ghstack-poisoned]

suo commented on torch/package/package_exporter.py:
```diff
@@ -454,6 +466,11 @@ def close(self):
+        # Check that all mock and extern modules with allow_empty=False were matched.
+        for i, (pattern, _, allow_empty) in enumerate(self.patterns):
+            if not allow_empty and i not in self.matched_patterns:
+                raise RuntimeError(f"Exporter did not match {pattern} to any modules")
```
lgtm, but let's define a new class of errors (EmptyMatchError?) so that users can catch it systematically.

suo replied:
And let's add some more explicit language tying this error to the choice of `allow_empty`, something like: we specified X pattern but it did not match any modules, this is an error because allow_empty=False.

SplitInfinity replied:
How should we print out `X pattern`? Right now it gets printed out using `_GlobGroup.__str__`, which produced something like `_GlobGroup(include=..., exclude=...)`. Would that make sense to someone using this?

Commit:
Update on "[package] Add `allow_empty` flag to mock and extern"


**Summary**
This commit adds an optional `allow_empty` argument to
`PackageExporter.mock` and `PackageExporter.extern` that allows certain
patterns for mocked modules and extern modules to be marked ones that
*must* be matched during the packaging process. If a mock or extern
module with `allow_empty=False` is not matched while packaging, an error
is thrown.

**Test Plan**
This commit adds two new test cases to `PackagingTest`,
`test_extern_glob_allow_empty` and `test_mock_glob_allow_empty` that
test this new flag. Existing tests already tests `allow_empty=True`.

**Fixes**
This commit fixes #53217.

[ghstack-poisoned]
